### ISSUE TITLE
Multi GPU, Multi node support

### ISSUE THREAD

**Issue by nakosung** [ibab/tensorflow-wavenet#140]
Is there anybody who wants wavenet to be distributed? I'm quite a newbie for tensorflow and have been investigating distributed features. 

This is my plan:
1. Step 1 - data parallelism (for faster learning)
  1. Single node / Multi GPU (following google's mnist-multigpu example)
  2. Multi node / Multi GPU (async)

2. Step 2 - model parallelism (for larger models)
 1. Single node / Multi GPU
 2. Multi node / Multi GPU

**Comment by nakosung** [comment]
Working on my fork: https://github.com/nakosung/tensorflow-wavenet/commit/1f00d7b72eeb93c4eff2d9ad664d956aef402457

**Comment by nakosung** [comment]
![image](https://cloud.githubusercontent.com/assets/2463571/19296698/a7009c84-9078-11e6-9e65-4e76cf0f1e09.png)

It seems to perform better.

**Comment by jyegerlehner** [comment]
> Is there anybody who wants wavenet to be distributed?

Yes please. I could use data parallelism.

I was thinking about context stacks, and I think the longer context wavenet and the shorter context wavenet could run on separate GPUs, as they have a relatively small interface (not much data to be transferred between GPUs). This might be a scenario for exploiting your #2 model parallelism.

**Comment by nakosung** [comment]
@jyegerlehner I think I could implement similar approach you described. :)

**Comment by nakosung** [comment]
For whom is interested, I've been working on distributed training on https://github.com/nakosung/tensorflow-wavenet/commit/a579a626b8182a638bfa5f35ebd996a526aeef60.

**Comment by nakosung** [comment]
Distributed training with data parallelism seems to work. I'll try to implement model parallelism after testing current implementation. (by spreading layers over nodes)

**Comment by nakosung** [comment]
Currently working on model parallelism. https://github.com/nakosung/tensorflow-wavenet/tree/model_parallel

**Comment by lemonzi** [comment]
If you want to merge data parallelism while you finish model parallelism (or even single-machine only), feel free to open a PR. It's always better to review small changes at a time.

**Comment by nakosung** [comment]
![image](https://cloud.githubusercontent.com/assets/2463571/19920557/c950a572-a11c-11e6-9f3f-812bf9c2a1a2.png)
Okay, multi-gpu seems working fine. I will try to make a PR for this feature. :)

**Comment by LeavesBreathe** [comment]
@nakosung I've been studying how you have done your model parallelism. 

Not that my design is much better, but it seems yours is rather inefficient. You are loading one "block(s)" on device 1 --> then the next series of blocks on device 2 and so on. 

This means that only one gpu is active at a time since block 1 must be computed before block 2 is calculated. Ideally, we would want each respective gpu to be 100% used all the time. 

Please let me know if I am misunderstanding the way your model is set up. I will try to experiment with multi gpu set ups to maximize memory and gpu usage time. I'll let you know if I get anything efficient.

**Comment by nakosung** [comment]
@LeavesBreathe Yes. My model parallelism implementation was not about 'speed'. It just distributes huge model into several pieces, but it turned out that this model was too slow. As far as I know, partitioning vertically would help model parallelism to be fast but it has some issues about *boundaries*.

**Comment by D-X-Y** [comment]
@nakosung  Hi, I'm also trying model parallelism too. I find some thing strange, that since I use a vgg-style cnn network, and when the whole network is on the same gpu, the gpu memory cost is 8GB, but when I seperate the network to put the first four layer on gpu 0 and the last on gpu 1. The total gpu memory cost becomes 10GB. It increases so much. Do you have any idea of that?

**Comment by nakosung** [comment]
@D-X-Y It seems tensorflow allocates some extra memory for distributed setting. (Maybe for recv-buffer for DMA?)
